components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java

```diff
@@ -622,13 +625,43 @@ public void confirmUserSelfRegistration(String code, String verifiedChannelType,
         HashMap<String, String> userClaims = getClaimsListToUpdate(user, recoveryData.getRemainingSetIds(),
                 externallyVerifiedClaim, recoveryData.getRecoveryScenario().toString());
 
+        if (RecoverySteps.VERIFY_EMAIL.equals(recoveryData.getRecoveryStep())) {
+            String pendingVerificationEmailClaimValue = getPendingVerificationEmailClaimValue(user, userStoreManager);
+            if (StringUtils.isNotBlank(pendingVerificationEmailClaimValue)) {
+                userClaims.put(IdentityRecoveryConstants.EMAIL_ADDRESS_PENDING_VALUE_CLAIM, "");
+                userClaims.put(IdentityRecoveryConstants.EMAIL_ADDRESS_CLAIM, pendingVerificationEmailClaimValue);
+                Utils.setThreadLocalToSkipSendingEmailVerificationOnUpdate(IdentityRecoveryConstants
+                        .SkipEmailVerificationOnUpdateStates.SKIP_ON_CONFIRM.toString());
+            }
+        }
+
         // Update the user claims.
         updateUserClaims(userStoreManager, user, userClaims);
 
         // Invalidate code.
         userRecoveryDataStore.invalidate(code);
     }
 
+    private String getPendingVerificationEmailClaimValue(User user, UserStoreManager
+            userStoreManager) throws IdentityRecoveryServerException {
+
+        Map<String, String> verificationPendingEmailClaimMap;
+        try {
+            verificationPendingEmailClaimMap = userStoreManager.getUserClaimValues(IdentityUtil.addDomainToName
+                    (user.getUserName(), user.getUserStoreDomain()), new String[]{IdentityRecoveryConstants
+                    .EMAIL_ADDRESS_PENDING_VALUE_CLAIM}, null);
```

Comment:
let's verify whether this works correctly when pointed to a JDBC userstore. Otherwise we need to use "default" instead of null.

Reply:
tested by pointing to h2 JDBC userstore. works as expected without any issue